### PR TITLE
Feature/14 review shiny ci cd

### LINKED ISSUE #14: Review Shiny CI/CD
https://mirai-solutions.ch/techguides/cicd-pipelines-for-automatic-deployment-of-a-r-shiny-web-app.html does not reflect the current reality: the content was written when Travis was the well-established tool and Actions was the not-yet-mature one, which is not quite the current situation.

A review is needed to have the primary focus on Actions.

### PR DESCRIPTION
**Pull request by Chargothrond**
Reviewed and updated shiny ci cd chapter, as well as the github actions yaml examples.
Closes #14.

### CODE REVIEW

File: shiny-ci-cd.Rmd [hunk 1]
```diff
@@ -34,9 +34,72 @@ Generally speaking, a CI/CD pipeline related to an R package is comprised of the
...
+[Workflows](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions) use `YAML` syntax and should be stored in the `.github/workflows` directory in the root of the repository.
+Workflows are constituted of jobs and each job is a set of steps to perform individual tasks, e.g. commands or actions.
+
+### Workflow steps
+
+A workflow should have an identifying `name` and an `on` section that indicates upon which events the workflow should be triggered. It should include at least one job and each job will have a set of steps fully specifying what to execute. Such steps can be a (predefined) action or a script, however, for the time being, GitHub Actions does not support a hierarchical, aggregated structure of actions. Being fully customizable, it is necessary to fully specify each step in the CI/CD pipeline explicitly.
+
+#### Setup
+
+- Checkout the source package from the repo, using `actions/checkout` provided by GitHub.
```
Comment (fvitalini) [inline shiny-ci-cd.Rmd:49]:
```suggestion
- Checkout the source package from the current repo, using the `actions/checkout` provided by GitHub.
```

File: shiny-ci-cd.Rmd [hunk 2]
```diff
@@ -34,9 +34,72 @@ Generally speaking, a CI/CD pipeline related to an R package is comprised of the
...
+- Setup R using action [`r-lib/actions/setup-r`](https://github.com/r-lib/actions#readme).
```
Comment (fvitalini) [inline shiny-ci-cd.Rmd:50]:
```suggestion
- Setup R using the action [`r-lib/actions/setup-r`](https://github.com/r-lib/actions#readme).
```

File: shiny-ci-cd.Rmd (hunk 1 above)
Comment (fvitalini) [reply shiny-ci-cd.Rmd:49]: If this step also includes ubuntu installation, we should mention it explicitly.

File: shiny-ci-cd.Rmd [hunk 3]
```diff
@@ -34,9 +34,72 @@ Generally speaking, a CI/CD pipeline related to an R package is comprised of the
...
+- Query and cache R package dependencies using `remotes::dev_package_deps()` and the `actions/cache` predefined action.
+- Install system dependencies using package `sysreqs::sysreq_commands()` (for the `ubuntu` runner used to run the action).
+- Install R package dependencies using `remotes::Install_deps()`.
+
+#### Package check
+
+- Check the package via using `rcmdcheck::rcmdcheck()`.
```
Comment (fvitalini) [inline shiny-ci-cd.Rmd:57]:
```suggestion
- Check the package via `rcmdcheck::rcmdcheck()`.
```

File: shiny-ci-cd.Rmd [hunk 4]
```diff
@@ -34,9 +34,72 @@ Generally speaking, a CI/CD pipeline related to an R package is comprised of the
...
+
+#### Deployment
+
+- Deploy to shinyapps.io, similar to the Travis CI approach:
```
Comment (fvitalini) [inline shiny-ci-cd.Rmd:61]:
```suggestion
- Deploy to shinyapps.io:
```
Reference to TravisCI is not needed anymore (also believe it is a legacy from when Travis was the default case upon which GitHub actions was building up)

File: shiny-ci-cd.Rmd [hunk 5]
```diff
@@ -34,9 +34,72 @@ Generally speaking, a CI/CD pipeline related to an R package is comprised of the
...
+  - An R script e.g. `deploy/deploy-shinyapps.R` (build-ignored via `usethis::use_build_ignore("deploy")`) defines the deployment commands based on the environment variables:
+```{r read-deploy-script, eval = TRUE, echo = FALSE}
+knitr::read_chunk("shiny-ci-cd/deploy/deploy-shinyapps.R", labels = "deploy-shinyapps")
+```
+```{r deploy-shinyapps}
+```
+
+### Workflow file
+
+The `steps` above are defined in the `.yml` workflow file as follows:
```
Comment (fvitalini) [inline shiny-ci-cd.Rmd:72]:
```suggestion
The `steps` described in the previous section are defined in the `.yml` workflow file as follows:
```

Comment (fvitalini) [reply shiny-ci-cd.Rmd:72]: not sure how to reference the section directly

File: shiny-ci-cd.Rmd [hunk 6]
```diff
@@ -34,9 +34,72 @@ Generally speaking, a CI/CD pipeline related to an R package is comprised of the
...
+
+```{r read-ci-cd.yml, eval = TRUE, include = FALSE}
+file.copy("shiny-ci-cd", "_book", recursive = TRUE)
+knitr::read_chunk("shiny-ci-cd/actions/ci-cd.yml", labels = "ci-cd.yml")
+```
+```{yml ci-cd.yml}
+```
+
+As visible from the run logs, all the CI/CD pipeline steps are performed subsequently, and are identifiable by the `name` field.
```
Comment (fvitalini) [inline shiny-ci-cd.Rmd:81]: You have never mentioned the run logs before, so if you want to mention them you should explain them I believe.

File: shiny-ci-cd.Rmd [hunk 7]
```diff
@@ -34,9 +34,72 @@ Generally speaking, a CI/CD pipeline related to an R package is comprised of the
...
 
+## GitHub Actions
+
+[GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions) is a service for running highly-customizable and flexible automated workflows, fully integrated with GitHub and very suitable to CI/CD pipelines.
+[Workflows](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions) use `YAML` syntax and should be stored in the `.github/workflows` directory in the root of the repository.
+Workflows are constituted of jobs and each job is a set of steps to perform individual tasks, e.g. commands or actions.
+
+### Workflow steps
+
+A workflow should have an identifying `name` and an `on` section that indicates upon which events the workflow should be triggered. It should include at least one job and each job will have a set of steps fully specifying what to execute. Such steps can be a (predefined) action or a script, however, for the time being, GitHub Actions does not support a hierarchical, aggregated structure of actions. Being fully customizable, it is necessary to fully specify each step in the CI/CD pipeline explicitly.
```
[the remark this answers is not in the document]
Comment (fvitalini) [reply shiny-ci-cd.Rmd:45]: Worth to mention somewhere that Actions in the workflow can be pulled from somewhere else and combined.

File: shiny-ci-cd.Rmd [hunk 8]
```diff
@@ -34,9 +34,72 @@ Generally speaking, a CI/CD pipeline related to an R package is comprised of the
...
+  # "https://mirai-solutions.ch/techguides/shiny-ci-cd/actions/ci-renv.yml"
+)
+usethis::use_github_actions_badge("CI-CD") # or "CI"
+```
+
+
 ## Travis CI
 
-Travis CI is an open-source continuous integration service that can be used to build and test software projects hosted on GitHub. To set up Travis CI you need to login at [https://travis-ci.com/](https://travis-ci.com/) (using your GitHub account) and provide authorization via GitHub (see [Travis CI Tutorial](https://docs.travis-ci.com/user/tutorial)).
+Travis CI is a previously fully open-source continuous integration service that can be used to build and test software projects hosted on GitHub. To set up Travis CI you need to login at [https://travis-ci.com/](https://travis-ci.com/) (using your GitHub account) and provide authorization via GitHub (see [Travis CI Tutorial](https://docs.travis-ci.com/user/tutorial)).
```
Comment (fvitalini) [inline shiny-ci-cd.Rmd:102]:
```suggestion
Travis CI is a continuous integration service, which used to be fully open-source, that can be used to build and test software projects hosted on GitHub. To set up Travis CI you need to login at [https://travis-ci.com/](https://travis-ci.com/) (using your GitHub account) and provide authorization via GitHub (see [Travis CI Tutorial](https://docs.travis-ci.com/user/tutorial)).
```

File: shiny-ci-cd.Rmd [hunk 9]
```diff
@@ -113,7 +173,6 @@ deploy:
     branch: master
```
Comment (fvitalini) [inline shiny-ci-cd.Rmd:173]: maybe somewhere we should add a note about the master branch being now called main if you create a new repo on GitHub

File: shiny-ci-cd/actions/ci-cd-renv.yml [hunk 10]
```diff
@@ -8,17 +8,32 @@ name: CI-CD
...
+    strategy:
+      # we keep a matrix for convenience, but we would always want to run on one
+      # single OS and R version
+      matrix:
+        config:
+          - {os: ubuntu-20.04, r: 'release'}
+
     env:
       RENV_PATHS_ROOT: ~/.local/share/renv
+      GITHUB_PAT: ${{ secrets.GITHUB_TOKEN }}
```
Comment (fvitalini) [inline shiny-ci-cd/actions/ci-cd-renv.yml:24]: shouldn't we add a comment here about this being the Personalized Access Token necessary to connect to GitHub?

File: shiny-ci-cd/actions/ci-cd-renv.yml [hunk 11]
```diff
@@ -8,17 +8,32 @@ name: CI-CD
 
 jobs:
   CI-CD:
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.config.os }}
```
Comment (fvitalini) [inline shiny-ci-cd/actions/ci-cd-renv.yml:11]: Can you add a comment here explaining how this differs from `ubuntu-latest`? It used to be a much clearer syntax

File: shiny-ci-cd/actions/ci-cd.yml [hunk 12]
```diff
@@ -6,46 +6,63 @@ name: CI-CD
 
 jobs:
   CI-CD:
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.config.os }}
+
+    name: ${{ matrix.config.os }} (${{ matrix.config.r }})
+
+    strategy:
+      # we keep a matrix for convenience, but we would always want to run on one
```
Comment (fvitalini) [inline shiny-ci-cd/actions/ci-cd.yml:14]: Why don't we have this comment and section also in the renv case?

File: shiny-ci-cd/actions/ci-cd.yml [hunk 13]
```diff
@@ -6,46 +6,63 @@ name: CI-CD
...
+      # single OS and R version
+      matrix:
+        config:
+          - {os: ubuntu-20.04, r: 'release', rspm: "https://packagemanager.rstudio.com/cran/__linux__/focal/latest"}
+
+    env:
+      RSPM: ${{ matrix.config.rspm }}
+      GITHUB_PAT: ${{ secrets.GITHUB_TOKEN }}
```
Comment (fvitalini) [inline shiny-ci-cd/actions/ci-cd.yml:22]: as in the renv case I'd add a comment here

File: shiny-ci-cd/actions/ci-cd.yml [hunk 14]
```diff
@@ -6,46 +6,63 @@ name: CI-CD
...
+          remotes::install_deps(dependencies = TRUE)
+          remotes::install_cran("rcmdcheck")
         shell: Rscript {0}
-        run: remotes::install_deps(dependencies = TRUE)
 
       - name: Check package
-        shell: Rscript {0}
         run: |
-          install.packages("rcmdcheck")
+          options(crayon.enabled = TRUE)
```
Comment (fvitalini) [inline shiny-ci-cd/actions/ci-cd.yml:63]: can you add a comment explaining why is this needed and what does it imply?

File: shiny-ci-cd/actions/ci-renv.yml [hunk 15]
```diff
@@ -8,17 +8,32 @@ name: CI-CD
 
 jobs:
   CI-CD:
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.config.os }}
+
+    name: ${{ matrix.config.os }} (${{ matrix.config.r }})
+
+    strategy:
```
Comment (fvitalini) [inline shiny-ci-cd/actions/ci-renv.yml:15]: same comments valid for the ci-cd case

File: shiny-ci-cd/actions/ci.yml [hunk 16]
```diff
@@ -6,43 +6,60 @@ name: CI-CD
 
 jobs:
   CI-CD:
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.config.os }}
+
```
Comment (fvitalini) [inline shiny-ci-cd/actions/ci.yml:10]: same comments as in the ci-cd case

**Review by fvitalini — CHANGES_REQUESTED** [review]
Some minor reqording. My main request is to clarify better some steps in the yml files by adding some comments.

File: shiny-ci-cd.Rmd (hunk 7 above)
[the remark this answers is not in the document]
Comment (Chargothrond) [reply shiny-ci-cd.Rmd]: Not sure what you mean by that. Can you either add some text yourself for this or elaborate / link to somewhere where this is explained?

File: shiny-ci-cd.Rmd [hunk 17]
```diff
@@ -34,9 +34,72 @@ Generally speaking, a CI/CD pipeline related to an R package is comprised of the
 
 Most of these steps are implemented by default in Travis CI for an R package. In GitHub Actions, on the other hand, it is currently necessary to manually specify each of them.
```
[the remark this answers is not in the document]
Comment (Chargothrond) [reply shiny-ci-cd.Rmd:35]: I disagree, the sentence above introduces the list of steps and then this sentence directly below it refers to those steps, so it should be clear what is meant and the order makes sense to me as well (i.e. mentioning how these steps relate to github actions and travis CI as a further, noteworthy detail).

File: shiny-ci-cd.Rmd (hunk 1 above)
Comment (Chargothrond) [reply shiny-ci-cd.Rmd:49]: It does not. I don't like the added article, will remove it on a later commit.

File: shiny-ci-cd.Rmd (hunk 2 above)
Comment (Chargothrond) [reply shiny-ci-cd.Rmd:50]: Here adding the article is fine, since there is also the noun "action" before the actual "name" / grammatical thing.

File: shiny-ci-cd/actions/ci-cd.yml (hunk 12 above)
Comment (Chargothrond) [reply shiny-ci-cd/actions/ci-cd.yml:14]: But we do? Either way, I will later align these sentences to what we changed in the ShinyCICD repo.

File: shiny-ci-cd/actions/ci-cd.yml (hunk 13 above)
Comment (Chargothrond) [reply shiny-ci-cd/actions/ci-cd.yml:22]: I don't see any comments regarding this section in any of the four .yml files in this folder, nor in the site.yaml. Not sure what you refer to.

As I already mentioned on https://github.com/miraisolutions/ShinyCICD/pull/1, we should view this together with the question there on how many / which comments we might want to add to these github actions workflows.

File: shiny-ci-cd/actions/ci-cd.yml (hunk 14 above)
Comment (Chargothrond) [reply shiny-ci-cd/actions/ci-cd.yml:63]: Same as on https://github.com/miraisolutions/ShinyCICD/pull/1, let's first discuss comments on these files in general.

File: shiny-ci-cd/actions/ci-cd-renv.yml (hunk 11 above)
Comment (Chargothrond) [reply shiny-ci-cd/actions/ci-cd-renv.yml:11]: Resolved as belongs to same topic of adding more comments to these files in general.

File: shiny-ci-cd/actions/ci-cd-renv.yml (hunk 10 above)
Comment (Chargothrond) [reply shiny-ci-cd/actions/ci-cd-renv.yml:24]: Resolved as belongs to same topic of adding more comments to these files in general.

File: shiny-ci-cd.Rmd (hunk 9 above)
Comment (Chargothrond) [reply shiny-ci-cd.Rmd:173]: Resolved as belongs to same topic of adding more comments to these files in general.

File: shiny-ci-cd.Rmd (hunk 6 above)
Comment (Chargothrond) [reply shiny-ci-cd.Rmd]: I updated this and it is more explicit now.

**Comment by Chargothrond** [comment]
Thanks for the feedback @fvitalini , integrated most of it apart from the comments regarding the .yml files.

As I mentioned already in https://github.com/miraisolutions/ShinyCICD/pull/1, we need to decide on how much we want to comment these files to explain them better.

@riccardoporreca Please let me know your opinion on this as well, and you can also review the changes now as I integrated Francesca's feedback.

File: shiny-ci-cd/actions/ci-cd.yml (hunk 13 above)
Comment (fvitalini) [reply shiny-ci-cd/actions/ci-cd.yml:22]: I meant the comment in the review

File: shiny-ci-cd.Rmd (hunk 7 above)
[the remark this answers is not in the document]
Comment (fvitalini) [reply shiny-ci-cd.Rmd]: I meant that one can take an action defined by someone and use it in his/her own workflow.

**Review by fvitalini — APPROVED** [review]

File: shiny-ci-cd.Rmd (hunk 7 above)
[the remark this answers is not in the document]
Comment (Chargothrond) [reply shiny-ci-cd.Rmd]: I see. I guess it is somewhat implied, but surely doesn't harm to point that out. I updated this with another commit.

File: .github/workflows/site.yaml [hunk 18]
```diff
@@ -11,8 +11,8 @@ jobs:
     name: ${{ matrix.config.os }} (${{ matrix.config.r }})
 
     strategy:
-      # we keep a matrix for convenience, but we would always want to run on one
-      # single OS and R version
+      # we keep a matrix for convenience, but we would typically just run on one
+      # single OS and R version, aligned with the target deployment environment
```
Comment (riccardoporreca) [inline .github/workflows/site.yaml:15]: This should be reverted, there is no target deployment environment for the website. This was the adaptation we did for the Shiny CI/CD workflows, which does not apply back to the techguides though.
```suggestion
      # we keep a matrix for convenience, but we would always want to run on one
      # single OS and R version
```

File: shiny-ci-cd.Rmd [hunk 19]
```diff
@@ -34,9 +34,72 @@ Generally speaking, a CI/CD pipeline related to an R package is comprised of the
...
+  # "https://mirai-solutions.ch/techguides/shiny-ci-cd/actions/ci-renv.yml"
+)
+usethis::use_github_actions_badge("CI-CD") # or "CI"
+```
+
+
 ## Travis CI
 
-Travis CI is an open-source continuous integration service that can be used to build and test software projects hosted on GitHub. To set up Travis CI you need to login at [https://travis-ci.com/](https://travis-ci.com/) (using your GitHub account) and provide authorization via GitHub (see [Travis CI Tutorial](https://docs.travis-ci.com/user/tutorial)).
+Travis CI is a continuous integration service, which used to be fully open-source, that can be used to build and test software projects hosted on GitHub. To set up Travis CI you need to login at [https://travis-ci.com/](https://travis-ci.com/) (using your GitHub account) and provide authorization via GitHub (see [Travis CI Tutorial](https://docs.travis-ci.com/user/tutorial)).
```
Comment (riccardoporreca) [inline shiny-ci-cd.Rmd:102]: The focus in not really on Travis no being open-source, but not really targeting / valuable for open-source projects.
```suggestion
Travis CI is a continuous integration service that can be used to build and test software projects hosted on GitHub. To set up Travis CI you need to login at [https://travis-ci.com/](https://travis-ci.com/) (using your GitHub account) and provide authorization via GitHub (see [Travis CI Tutorial](https://docs.travis-ci.com/user/tutorial)).

Travis CI used to be a very established, mature and popular tool in the open-source source community, before a recent change of policy made it less focused on open-source, offering only limited free trial plans.
```

File: shiny-ci-cd.Rmd [hunk 20]
```diff
@@ -34,9 +34,72 @@ Generally speaking, a CI/CD pipeline related to an R package is comprised of the
...
 
+## GitHub Actions
+
+[GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions) is a service for running highly-customizable and flexible automated workflows, fully integrated with GitHub and very suitable to CI/CD pipelines.
+[Workflows](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions) use `YAML` syntax and should be stored in the `.github/workflows` directory in the root of the repository.
+Workflows are constituted of jobs and each job is a set of steps to perform individual tasks, e.g. commands or actions.
+
+### Workflow steps
+
+A workflow should have an identifying `name` and an `on` section that indicates upon which events the workflow should be triggered. It should include at least one job and each job will have a set of steps fully specifying what to execute. Such steps can be an action (predefined, sourcing from GitHub repos that contain such actions) or a script. However, for the time being, GitHub Actions does not support a hierarchical, aggregated structure of actions. Being entirely customizable, it is necessary to fully specify each step in the CI/CD pipeline explicitly.
```
Comment (riccardoporreca) [inline shiny-ci-cd.Rmd:45]: There is something going on by now, although not mature yet
```suggestion
A workflow should have an identifying `name` and an `on` section that indicates upon which events the workflow should be triggered. It should include at least one job and each job will have a set of steps fully specifying what to execute. Such steps can be an action (predefined, sourcing from GitHub repos that contain such actions) or a script. However, for the time being, GitHub Actions does not provide a mature support for hierarchical, aggregated structure of actions. Being entirely customizable, it is necessary to fully specify each step in the CI/CD pipeline explicitly.
```

File: shiny-ci-cd.Rmd [hunk 21]
```diff
@@ -34,9 +34,72 @@ Generally speaking, a CI/CD pipeline related to an R package is comprised of the
...
+
+#### Setup
+
+- Checkout the source package from the current repo, using `actions/checkout` provided by GitHub.
```
Comment (riccardoporreca) [inline shiny-ci-cd.Rmd:49]: We are talking about "the repository" in the intro, there is no "current repository" for a techguides reader 
```suggestion
- Checkout the source package from the repository, using `actions/checkout` provided by GitHub.
```

File: shiny-ci-cd.Rmd [hunk 22]
```diff
@@ -34,9 +34,72 @@ Generally speaking, a CI/CD pipeline related to an R package is comprised of the
...
+- Setup R using the action [`r-lib/actions/setup-r`](https://github.com/r-lib/actions#readme).
+- Query and cache R package dependencies using `remotes::dev_package_deps()` and the `actions/cache` predefined action.
+- Install system dependencies using package `sysreqs::sysreq_commands()` (for the `ubuntu` runner used to run the action).
```
Comment (riccardoporreca) [inline shiny-ci-cd.Rmd:52]:
```suggestion
- Install system dependencies (for the `ubuntu` runner defined for the workflow) using `sysreqs::sysreq_commands()`.
```

File: shiny-ci-cd.Rmd [hunk 23]
```diff
@@ -34,9 +34,72 @@ Generally speaking, a CI/CD pipeline related to an R package is comprised of the
...
+- Install R package dependencies using `remotes::Install_deps()`.
```
Comment (riccardoporreca) [inline shiny-ci-cd.Rmd:53]:
```suggestion
- Install R package dependencies using `remotes::install_deps()`.
```

File: shiny-ci-cd.Rmd [hunk 24]
```diff
@@ -34,9 +34,72 @@ Generally speaking, a CI/CD pipeline related to an R package is comprised of the
...
+
+#### Package check
+
+- Check the package via `rcmdcheck::rcmdcheck()`.
+
+#### Deployment
+
+- Deploy to shinyapps.io:
+  - In this case, environment variables `SHINYAPPS_ACCOUNT`, `SHINYAPPS_TOKEN` and `SHINYAPPS_SECRET` (defining credentials for [shinyapps.io](https://www.shinyapps.io/)), are specified / accessible as GitHub [secrets](https://docs.github.com/en/free-pro-team@latest/actions/reference#authentication-and-secrets).
```
Comment (riccardoporreca) [inline shiny-ci-cd.Rmd:62]: We need to introduce this better, as was originally done with the Travis as the first part of the guide
```suggestion
  - In order to provide credentials for the deployment, account name and corresponding [tokens](https://docs.rstudio.com/shinyapps.io/getting-started.html#deploying-applications) for shinyapps.io, environment variables `SHINYAPPS_ACCOUNT`, `SHINYAPPS_TOKEN` and `SHINYAPPS_SECRET` are specified / accessible as GitHub [secrets](https://docs.github.com/en/free-pro-team@latest/actions/reference#authentication-and-secrets).
```

File: shiny-ci-cd.Rmd [hunk 25]
```diff
@@ -34,9 +34,72 @@ Generally speaking, a CI/CD pipeline related to an R package is comprised of the
...
+The `steps` described in the previous section are defined in the `.yml` workflow file as follows:
+
+```{r read-ci-cd.yml, eval = TRUE, include = FALSE}
+file.copy("shiny-ci-cd", "_book", recursive = TRUE)
+knitr::read_chunk("shiny-ci-cd/actions/ci-cd.yml", labels = "ci-cd.yml")
+```
+```{yml ci-cd.yml}
+```
+
+As visible from the run logs that can be found on GitHub under the `Actions` tab, all the CI/CD pipeline steps are performed subsequently, and are identifiable by the `name` field. See the example below:
```
Comment (riccardoporreca) [inline shiny-ci-cd.Rmd:81]:
```suggestion
As visible from the run logs that can be found in the GitHub repository under the `Actions` tab, all the CI/CD pipeline steps are performed subsequently, and are identifiable by the `name` field. See the example below:
```

File: shiny-ci-cd.Rmd [hunk 26]
```diff
@@ -34,9 +34,72 @@ Generally speaking, a CI/CD pipeline related to an R package is comprised of the
...
+
+![GitHub Actions Continuous Integration / Continuous Deployment pipeline for a packaged Shiny app](shiny-ci-cd/img/ShinyCICD_githubactions1.png)
```
Comment (riccardoporreca) [inline shiny-ci-cd.Rmd:83]: We need a new image now to be aligned with the steps as revised

File: shiny-ci-cd.Rmd [hunk 27]
```diff
@@ -34,9 +34,72 @@ Generally speaking, a CI/CD pipeline related to an R package is comprised of the
 
 Most of these steps are implemented by default in Travis CI for an R package. In GitHub Actions, on the other hand, it is currently necessary to manually specify each of them.
 
+## GitHub Actions
+
+[GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions) is a service for running highly-customizable and flexible automated workflows, fully integrated with GitHub and very suitable to CI/CD pipelines.
+[Workflows](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions) use `YAML` syntax and should be stored in the `.github/workflows` directory in the root of the repository.
+Workflows are constituted of jobs and each job is a set of steps to perform individual tasks, e.g. commands or actions.
+
```
Comment (riccardoporreca) [inline shiny-ci-cd.Rmd:42]:
```suggestion
Workflows are constituted of jobs and each job is a set of steps to perform individual tasks, e.g. commands or actions.

The next sections describe in detail the relevant workflow steps of a typical CI/CD pipeline for a packages Shiny app, also covering the usage of `renv` to track package dependencies. Finally, we will show how you can use the convenience function `usethis::use_github_action()` for including such workflows in you project.
```

File: shiny-ci-cd.Rmd [hunk 28]
```diff
@@ -34,9 +34,72 @@ Generally speaking, a CI/CD pipeline related to an R package is comprised of the
...
+file.copy("shiny-ci-cd", "_book", recursive = TRUE)
+knitr::read_chunk("shiny-ci-cd/actions/ci-cd.yml", labels = "ci-cd.yml")
+```
+```{yml ci-cd.yml}
+```
+
+As visible from the run logs that can be found on GitHub under the `Actions` tab, all the CI/CD pipeline steps are performed subsequently, and are identifiable by the `name` field. See the example below:
+
+![GitHub Actions Continuous Integration / Continuous Deployment pipeline for a packaged Shiny app](shiny-ci-cd/img/ShinyCICD_githubactions1.png)
+
```
Comment (riccardoporreca) [inline shiny-ci-cd.Rmd:84]: We completely miss the part about `renv`, which is very relevant to us. We can introduce this new section in the very same way as currently done in the Travis part (perhaps making that part slimmer do avoid redundancy), and highlight which steps are different (caching and restoring locked dependencies). Not sure we want to show the full worfklow, maybe just the `ci-renv.yml`, since deployment is unaffected

File: shiny-ci-cd.Rmd [hunk 29]
```diff
@@ -34,9 +34,72 @@ Generally speaking, a CI/CD pipeline related to an R package is comprised of the
...
+### `usethis::use_github_action()`
+
+In order to use the GitHub action workflow above, or its renv-based variant, you can simply:
```
Comment (riccardoporreca) [inline shiny-ci-cd.Rmd:87]:
```suggestion
In order to setup and use CI/CD GitHub Actions workflows as described above, we provide full YAML workflow files for CI and CI/CD pipelines with and without `renv`, which you can simply include in your project via:
```

**Review by riccardoporreca — CHANGES_REQUESTED** [review]
See some refinements following the main restructuring with Actions are being the prominent part here. The main gap is talking about renv in the actions section.

File: shiny-ci-cd.Rmd (hunk 21 above)
Comment (Chargothrond) [reply shiny-ci-cd.Rmd:47]: we are not consistent with repository / repo in all our articles (perhaps not in the techguides, just remember seeing both being used). I would favor repository for our techguides and articles.

File: shiny-ci-cd.Rmd (hunk 26 above)
Comment (Chargothrond) [reply shiny-ci-cd.Rmd:82]: I could take a screenshot from https://github.com/miraisolutions/ShinyCICD/runs/2508242565. For some reason, the title changed from CI-CD / CI-CD to ubuntu-20.04 (release). Let me know if we should investigate this or just assume that something changed in the background on how this title is derived and that it is fine to keep the new name.

File: shiny-ci-cd.Rmd (hunk 28 above)
Comment (Chargothrond) [reply shiny-ci-cd.Rmd:83]: This requires some thinking, as including the full ci-cd-renv.yml file just after ci-cd.yml seems a bit excessive (same with including a new screenshot for the renv steps). We could use collapsible sections for this or even put it there entirely, not sure what looks more appealing.

File: shiny-ci-cd.Rmd (hunk 26 above)
Comment (riccardoporreca) [reply shiny-ci-cd.Rmd:82]: I agree that a screenshot from the run you propose will do.

> the title changed from CI-CD / CI-CD to ubuntu-20.04 (release)

This is due to the fact that we set the `job`  `name: ` to the matrix entries for OS and R version. This makes sense and I think it is OK. CI-CD is also the name of the workflow and is now shown in the left side bar and not in the title of the run  output, which now only contains the job name.

File: shiny-ci-cd.Rmd (hunk 28 above)
Comment (riccardoporreca) [reply shiny-ci-cd.Rmd:83]: I suggest to go ahead with the introduction, just mentionthat `renv` requires a different caching and that dependencies are installed via `renv::restore`, pointing to https://rstudio.github.io/renv/articles/ci.html#github-actions.

I would not include new screenshots here, and would avoid the the full workflow in the `renv` section. Still, I think it makes sense to perhaps enhance the section about `usethis` to give more relevance to the 4 full workflow .yml files we provide and show them, indeed using collapsible / accordion. We could do this in R markdown using bsplus (http://ijlyttle.github.io/bsplus/index.html), see how tit was done for the CovidR contest in  https://github.com/Milano-R/erum2020-covidr-contest/blob/a13b0ee2b2eec57bc523c7d6e50734863e35a9d8/contest/index.Rmd#L86-L103

Overall we could have stg like 

```md

### Complete workflows and `usethis::use_github_action()`

Full YAML workflows for CI and CI/CD pipelines, with and without `renv` are shown below and provided as part of this guide.

In order to setup and use CI/CD GitHub Actions workflows as described above, you can simply include the relevant workflow file your project via:

[...] existing code snippet

[...] collapsible panel for each workflow files (perhaps within a dedicated `####` sub section), the panel (and subsection) title being the file name and the content is the file content

File: shiny-ci-cd.Rmd (hunk 26 above)
Comment (Chargothrond) [reply shiny-ci-cd.Rmd:83]: updated and pushed

File: shiny-ci-cd.Rmd (hunk 28 above)
Comment (Chargothrond) [reply shiny-ci-cd.Rmd:84]: @riccardoporreca I updated part of this, but could not add collapsible sections using bsplus (nor other approaches I tried from some quick googling + trying out) - it seems that collapsible sections don't work with bookdown and / or gitbook.

Before I do a more thorough investigation, would you know something about this? Maybe some work-around with html and possibly other things? I was searching for something about "collapsible sections" in the bookdown documentation, but could not find anything helpful regarding this problem.

Comment (riccardoporreca) [reply shiny-ci-cd.Rmd:92]: - Collapsible full workflow files included using the `<details>` (https://www.w3schools.com/tags/tag_details.asp) in commit 19b455d 
- Explicit section about renv under GitHub action included as hinted above in commit 900444e

**Comment by riccardoporreca** [comment]
@Chargothrond, see my follow-up updates and other smaller refinements I committed.

We may want to make sure there is full alignment with the workflows as showcased in https://github.com/miraisolutions/ShinyCICD/pull/1

**Review by riccardoporreca — CHANGES_REQUESTED** [review]
We should also delete file `.travis.yml`, as Travis will not run any more and the focus is now fully on GitHub Actions

**Comment by Chargothrond** [comment]
@riccardoporreca I have aligned the site.yml and deleted the travis.yml, let me know if anything else is required, otherwise you can go ahead with merging.

File: shiny-ci-cd/actions/ci-cd-renv.yml [hunk 30]
```diff
@@ -27,25 +44,27 @@ jobs:
...
-        env:
-          RHUB_PLATFORM: linux-x86_64-ubuntu-gcc
         run: |
-          Rscript -e "renv::install('r-hub/sysreqs')"
-          sysreqs=$(Rscript -e "cat(sysreqs::sysreq_commands('DESCRIPTION'))")
-          sudo -s eval "$sysreqs"
+          while read -r cmd
+          do
+            eval sudo $cmd
+          done < <(Rscript -e 'writeLines(remotes::system_requirements("ubuntu", "20.04"))')
```
Comment (riccardoporreca) [inline shiny-ci-cd/actions/ci-cd-renv.yml:51]: In general, we there is no guarantee of having `remotes` installed
```suggestion
          Rscript -e "install.packages('remotes')"
          while read -r cmd
          do
            eval sudo $cmd
          done < <(Rscript -e 'writeLines(remotes::system_requirements("ubuntu", "20.04"))')
```

File: shiny-ci-cd/actions/ci-renv.yml [hunk 31]
```diff
@@ -27,19 +44,20 @@ jobs:
...
-        env:
-          RHUB_PLATFORM: linux-x86_64-ubuntu-gcc
         run: |
-          Rscript -e "renv::install('r-hub/sysreqs')"
-          sysreqs=$(Rscript -e "cat(sysreqs::sysreq_commands('DESCRIPTION'))")
-          sudo -s eval "$sysreqs"
+          while read -r cmd
+          do
+            eval sudo $cmd
+          done < <(Rscript -e 'writeLines(remotes::system_requirements("ubuntu", "20.04"))')
```
Comment (riccardoporreca) [inline shiny-ci-cd/actions/ci-renv.yml:51]: See above
```suggestion
          Rscript -e "install.packages('remotes')"
          while read -r cmd
          do
            eval sudo $cmd
          done < <(Rscript -e 'writeLines(remotes::system_requirements("ubuntu", "20.04"))')
```

**Review by riccardoporreca — APPROVED** [review]
Refined comment about renv and fixed installation of `remotes` with `renv` as a final review.

I will commit the suggestions and squash-merge.
